File: integrations/sysforge/static/js/views/diagnostics.js
/**
 * Business Diagnostics — read-only live DB snapshot.
 * Product panel only (strip agent debug-session file writers — chat b89af01b).
 * Diagnostics list ≠ search proof (BUG-018).
 */

let _abort = null;
let _busy = false;

/**
 * @param {HTMLElement} container
 * @param {{ api: (path: string, opts?: object) => Promise<object> }} deps
 */
export function mountDiagnostics(container, deps) {
  if (container.dataset.mounted === '1') return;

  container.innerHTML = `
    <div class="sysforge-diagnostics">
      <div class="sysforge-diagnostics-header">
        <div>
          <h3 class="sysforge-view-heading" tabindex="-1">Business Diagnostics</h3>
          <p class="sysforge-dashboard-lead">
            Live snapshot of what is stored in the Business database.
          </p>
        </div>
        <button type="button" class="btn-primary" id="sysforge-diagnostics-refresh">
          Refresh
        </button>
      </div>
      <p class="sysforge-diagnostics-error" id="sysforge-diagnostics-error" hidden></p>
      <dl class="sysforge-diagnostics-summary" id="sysforge-diagnostics-summary">
        <div><dt>Health</dt><dd id="sysforge-diag-health">—</dd></div>
        <div><dt>Path</dt><dd id="sysforge-diag-path">—</dd></div>
        <div><dt>Size</dt><dd id="sysforge-diag-size">—</dd></div>
        <div><dt>Latest migration</dt><dd id="sysforge-diag-migration">—</dd></div>
        <div><dt>Applied</dt><dd id="sysforge-diag-applied">—</dd></div>
        <div><dt>Clients</dt><dd id="sysforge-diag-clients">—</dd></div>
        <div><dt>Invoices</dt><dd id="sysforge-diag-invoices">—</dd></div>
        <div><dt>Parts</dt><dd id="sysforge-diag-parts">—</dd></div>
      </dl>
      <p class="sysforge-diagnostics-message" id="sysforge-diagnostics-message" hidden></p>
      <p class="sysforge-diagnostics-note" id="sysforge-diagnostics-note">
        Showing rows here means they are in SQLite. Client search may still miss
        them until search reindex runs after import/restore.
      </p>
    </div>`;

  const refreshBtn = container.querySelector('#sysforge-diagnostics-refresh');

  function setBusy(on) {
    _busy = on;
    if (refreshBtn) refreshBtn.disabled = on;
  }

  function showError(msg) {
    const el = container.querySelector('#sysforge-diagnostics-error');
    if (!el) return;
    if (msg) {
      el.textContent = msg;
      el.hidden = false;
    } else {
      el.textContent = '';
      el.hidden = true;
    }
  }

  function applySnapshot(data) {
    const db = data.db || {};
    const schema = data.schema || {};
    const counts = data.counts || {};
    const setText = (id, value) => {
      const el = container.querySelector(id);
      if (el) el.textContent = value == null || value === '' ? '—' : String(value);
    };
    setText('#sysforge-diag-health', data.health || '—');
    setText('#sysforge-diag-path', db.path_display || 'plugins/sysforge/sysforge.db');
    setText('#sysforge-diag-size', db.size_display || '(missing)');
    setText(
      '#sysforge-diag-migration',
      schema.latest_migration_id || schema.latest_migration_name || '(none)'
    );
    setText(
      '#sysforge-diag-applied',
      schema.applied_count != null ? String(schema.applied_count) : '—'
    );
    setText('#sysforge-diag-clients', counts.clients != null ? String(counts.clients) : '—');
    setText('#sysforge-diag-invoices', counts.invoices != null ? String(counts.invoices) : '—');
    setText(
      '#sysforge-diag-parts',
      counts.parts == null ? '—' : String(counts.parts)
    );

    const msgEl = container.querySelector('#sysforge-diagnostics-message');
    if (msgEl) {
      if (data.message) {
        msgEl.textContent = data.message;
        msgEl.hidden = false;
      } else {
        msgEl.textContent = '';
        msgEl.hidden = true;
      }
    }
    const noteEl = container.querySelector('#sysforge-diagnostics-note');
    if (noteEl && data.note) {
      noteEl.textContent = data.note;
    }
  }

  async function reload() {
    if (_busy) return;
    showError('');
    if (_abort) _abort.abort();
    _abort = typeof AbortController !== 'undefined' ? new AbortController() : null;
    setBusy(true);
    try {
      const data = await deps.api('/diagnostics', {
        signal: _abort?.signal,
      });
      if (_abort?.signal?.aborted) return;
      applySnapshot(data || {});
    } catch (err) {
      if (err?.name === 'AbortError') return;
      showError(err?.message || String(err));
    } finally {
      setBusy(false);
    }
  }

  refreshBtn?.addEventListener('click', () => {
    reload().catch(() => {});
  });

  container.dataset.mounted = '1';
  container._sysforgeReloadDiagnostics = reload;
  container._sysforgeAbortDiagnostics = () => {
    if (_abort) _abort.abort();
    _abort = null;
  };
}

/**
 * @param {HTMLElement} container
 */
export async function activateDiagnostics(container) {
  const heading = container.querySelector('.sysforge-view-heading');
  heading?.focus?.({ preventScroll: true });
  if (typeof container._sysforgeReloadDiagnostics === 'function') {
    try {
      await container._sysforgeReloadDiagnostics();
    } catch (_) {
      /* errors rendered in panel */
    }
  }
}

/**
 * @param {HTMLElement} container
 */
export function deactivateDiagnostics(container) {
  if (typeof container._sysforgeAbortDiagnostics === 'function') {
    container._sysforgeAbortDiagnostics();
  }
}

export default { mountDiagnostics, activateDiagnostics, deactivateDiagnostics };

File: integrations/sysforge/diagnostics.py
"""Read-only Business DB diagnostics snapshot.

Product Diagnostics panel only — no DebugLog / NDJSON / .cursor/debug.log writers
(chat b89af01b). Absolute paths stay redacted unless the caller is admin and
passes reveal_path=1.
"""

from __future__ import annotations

import sqlite3
from pathlib import Path
from typing import Any

from integrations.sysforge.db import connection as db_connection

PATH_DISPLAY = "plugins/sysforge/sysforge.db"
PLUGIN_ID = "sysforge"


def _format_size(size_bytes: int) -> str:
    if size_bytes < 1024:
        return f"{size_bytes} B"
    kb = size_bytes / 1024.0
    if kb < 1024:
        return f"{kb:.1f} KB"
    mb = kb / 1024.0
    return f"{mb:.1f} MB"


def _table_exists(conn: sqlite3.Connection, name: str) -> bool:
    row = conn.execute(
        "SELECT 1 FROM sqlite_master WHERE type='table' AND name=? LIMIT 1",
        (name,),
    ).fetchone()
    return row is not None


def _safe_count(conn: sqlite3.Connection, sql: str) -> int | None:
    try:
        row = conn.execute(sql).fetchone()
        return int(row[0]) if row else 0
    except sqlite3.Error:
        return None


def collect_diagnostics(
    *,
    db_path: Path | None = None,
    reveal_path: bool = False,
) -> dict[str, Any]:
    """Build the diagnostics DTO. Never reads config secrets or host env."""
    target = db_path if db_path is not None else db_connection.db_path()
    result: dict[str, Any] = {
        "ok": True,
        "plugin_id": PLUGIN_ID,
        "db": {
            "exists": False,
            "readable": False,
            "size_bytes": 0,
            "size_display": "(missing)",
            "path_display": PATH_DISPLAY,
            "path_absolute": str(target.resolve()) if reveal_path else None,
        },
        "schema": {
            "latest_migration_id": None,
            "latest_migration_name": None,
            "applied_count": 0,
            "applied": [],
        },
        "counts": {
            "clients": 0,
            "invoices": 0,
            "parts": None,
        },
        "health": "missing",
        "message": None,
        # Diagnostics lists rows in SQLite; client search may still miss them
        # until FTS reindex after import/restore (BUG-018).
        "note": (
            "Showing rows here means they are in SQLite. "
            "Client search may still miss them until search reindex runs "
            "after import/restore."
        ),
    }

    if not target.is_file():
        return result

    result["db"]["exists"] = True
    try:
        size = target.stat().st_size
        result["db"]["size_bytes"] = size
        result["db"]["size_display"] = _format_size(size)
    except OSError as exc:
        result["health"] = "error"
        result["message"] = f"Cannot stat database: {exc}"
        return result

    conn: sqlite3.Connection | None = None
    try:
        # Read-only open; do not create the file if missing (already checked).
        uri = target.resolve().as_uri() + "?mode=ro"
        conn = sqlite3.connect(uri, uri=True, check_same_thread=False)
        conn.row_factory = sqlite3.Row
        result["db"]["readable"] = True

        if not _table_exists(conn, "SchemaVersion"):
            result["health"] = "schema_pending"
            result["message"] = "Migrations not applied."
            return result

        rows = conn.execute(
            "SELECT Id, Name, AppliedAt FROM SchemaVersion ORDER BY Id"
        ).fetchall()
        applied = [
            {
                "id": int(r[0]),
                "name": str(r[1]),
                "applied_at": str(r[2]) if r[2] is not None else None,
            }
            for r in rows
        ]
        result["schema"]["applied"] = applied
        result["schema"]["applied_count"] = len(applied)
        if applied:
            latest = applied[-1]
            # Prefer zero-padded file stem style for UI (e.g. 0016_clients_fts).
            result["schema"]["latest_migration_id"] = latest["name"]
            result["schema"]["latest_migration_name"] = latest["name"]

        if _table_exists(conn, "Clients"):
            result["counts"]["clients"] = _safe_count(
                conn, "SELECT COUNT(*) FROM Clients WHERE IsDeleted = 0"
            ) or 0
        if _table_exists(conn, "Invoices"):
            # Include incomplete / all statuses (desktop GetInvoices includeIncomplete).
            result["counts"]["invoices"] = _safe_count(
                conn, "SELECT COUNT(*) FROM Invoices"
            ) or 0
        if _table_exists(conn, "Parts"):
            result["counts"]["parts"] = _safe_count(
                conn, "SELECT COUNT(*) FROM Parts"
            )

        # Quick integrity hint (non-blocking for panel; errors degrade health).
        try:
            check = conn.execute("PRAGMA integrity_check").fetchone()
            ok_text = str(check[0]).lower() if check else ""
            if ok_text != "ok":
                result["health"] = "error"
                result["message"] = f"Integrity check: {check[0] if check else 'unknown'}"
                return result
        except sqlite3.Error as exc:
            result["health"] = "error"
            result["message"] = f"Integrity check failed: {exc}"
            return result

        if result["schema"]["applied_count"] == 0:
            result["health"] = "schema_pending"
            result["message"] = "Migrations not applied."
        else:
            result["health"] = "ok"
        return result
    except sqlite3.Error as exc:
        result["health"] = "error"
        result["message"] = f"Database error: {exc}"
        return result
    finally:
        if conn is not None:
            conn.close()
